core/src/main/scala/cats/Applicative.scala

```diff
@@ -92,18 +94,18 @@ trait Applicative[F[_]] extends Apply[F] with InvariantMonoidal[F] { self =>
       val one = map(fa)(Chain.one(_))
 
       // invariant: n >= 1
-      def loop(n: Int): F[Chain[A]] =
-        if (n == 1) one
-        else {
+      @tailrec def loop(fa: F[Chain[A]], n: Int, acc: F[Chain[A]]): F[Chain[A]] =
+        if (n == 1) map2(fa, acc)(_.concat(_))
+        else
           // n >= 2
           // so (n >> 1) >= 1 and we are allowed to call loop
-          val half = loop(n >> 1)
-          val both = map2(half, half)(_.concat(_))
-          if ((n & 1) == 1) map2(one, both)(_.concat(_))
-          else both
-        }
+          loop(
+            map2(fa, fa)(_.concat(_)),
+            n >> 1,
+            if ((n & 1) == 1) map2(acc, fa)(_.concat(_)) else acc
+          )
 
-      map(loop(n))(_.toList)
+      map(loop(one, n, pure(Chain.empty)))(_.toList)
```

Comment:
I have a minor anxiety about the extra cost of the `map2(a, pure(Chain.empty))(_.concat(_))` being the same as `a`.

What if we cheat:
```scala
      val emptyF = pure(Chain.empty)

      def concat(left: F[Chain[A]], right: F[Chain[A]]): F[Chain[A]] =
        if (right eq emptyF) left else map2(left, right)(_.concat(_))

      @tailrec def loop(fa: F[Chain[A]], n: Int, acc: F[Chain[A]]): F[Chain[A]] =
         if (n == 1) concat(fa, acc)
         else
           // n >= 2
           // so (n >> 1) >= 1 and we are allowed to call loop
           loop(
             map2(fa, fa)(_.concat(_)),
             n >> 1,
             if ((n & 1) == 1) concat(fa, acc) else acc
           )
```
So we can leverage the law that `concat(a, pure(Chain.empty)) == a`

Reply:
What if we wrote it as `loop(one, n - 1, one)` instead?

Reply:
yeah, I think that works... but for `replicateA_` I think we need to rewrite so that we only entire the loop in the `n >= 2` case as we do here.